src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TryStatement.cs

```diff
@@ -16,13 +16,60 @@ public override BoundNode VisitTryStatement(BoundTryStatement node)
             var origSawAwait = _sawAwait;
             _sawAwait = false;
 
-            ImmutableArray<BoundCatchBlock> catchBlocks = (ImmutableArray<BoundCatchBlock>)this.VisitList(node.CatchBlocks);
+            var optimizing = this._compilation.Options.OptimizationLevel == OptimizationLevel.Release;
+            ImmutableArray<BoundCatchBlock> catchBlocks =
+                // When optimizing and we have a morally empty try block, we can discard the catch blocks.
+                (optimizing && IsMorallyEmpty(tryBlock)) ? ImmutableArray<BoundCatchBlock>.Empty
+                : this.VisitList(node.CatchBlocks);
             BoundBlock finallyBlockOpt = (BoundBlock)this.Visit(node.FinallyBlockOpt);
 
             _sawAwaitInExceptionHandler |= _sawAwait;
             _sawAwait |= origSawAwait;
 
-            return node.Update(tryBlock, catchBlocks, finallyBlockOpt, node.PreferFaultHandler);
+            if (optimizing && IsMorallyEmpty(finallyBlockOpt))
+            {
+                finallyBlockOpt = null;
+            }
+
+            return (catchBlocks.IsDefaultOrEmpty && finallyBlockOpt == null)
+                ? (BoundNode)tryBlock
+                : (BoundNode)node.Update(tryBlock, catchBlocks, finallyBlockOpt, node.PreferFaultHandler);
+        }
+
+        /// <summary>
+        /// Is there no code to execute in the given statement that could have side-effects,
+        /// such as throwing an exception?
+        /// </summary>
+        private static bool IsMorallyEmpty(BoundStatement statement)
```

Comment:
Agreed. Consider a name such as HasNoSideEffects and update comments in callers.